File: src/commands/validate-schemas.ts
import { Command } from 'commander';
import chalk from 'chalk';
import { validateSchemaCoverage, writeSchemaValidationReport } from '../diff.js';

export const validateSchemasCommand = new Command('validate-schemas')
  .description('Validate schema coverage in an OpenAPI spec')
  .requiredOption('-s, --spec <path>', 'path to OpenAPI spec')
  .option('-o, --output <path>', 'path to write JSON validation report')
  .option('--exit-on-gap', 'exit with code 2 if gaps are found')
  .action((options: { spec: string; output?: string; exitOnGap?: boolean }) => {
    const result = validateSchemaCoverage(options.spec);

    console.info('\n' + chalk.bold('Schema Coverage Report:'));
    console.info(`  Total endpoints: ${result.summary.totalEndpoints}`);
    console.info(
      `  Missing response schemas: ${result.summary.missingResponseSchemas > 0 ? chalk.red(String(result.summary.missingResponseSchemas)) : chalk.green('0')}`
    );
    console.info(
      `  Missing request schemas: ${result.summary.missingRequestSchemas > 0 ? chalk.red(String(result.summary.missingRequestSchemas)) : chalk.green('0')}`
    );
    console.info(
      `  Bare response schemas: ${result.summary.bareResponseSchemas > 0 ? chalk.yellow(String(result.summary.bareResponseSchemas)) : chalk.green('0')}`
    );
    console.info(
      `  Missing parameter schemas: ${result.summary.missingParamSchemas > 0 ? chalk.red(String(result.summary.missingParamSchemas)) : chalk.green('0')}`
    );
    console.info(`  Total gaps: ${result.summary.totalGaps > 0 ? chalk.red(String(result.summary.totalGaps)) : chalk.green('0')}`);

    if (result.gaps.length > 0) {
      const byCategory = new Map<string, SchemaGap[]>();
      for (const gap of result.gaps) {
        const list = byCategory.get(gap.category) || [];
        list.push(gap);
        byCategory.set(gap.category, list);
      }

      for (const [category, gaps] of byCategory) {
        console.info('\n' + chalk.yellow(`${category}:`));
        for (const gap of gaps) {
          console.info(chalk.yellow(`  ${gap.method} ${gap.path} | ${gap.operationId}`));
          console.info(chalk.gray(`    → ${gap.detail}`));
        }
      }
    } else {
      console.info('\n' + chalk.green('✓ All endpoints have complete schema coverage!'));
    }

    if (options.output) {
      writeSchemaValidationReport(result, options.output);
      console.info('\n' + chalk.green(`Report written to ${options.output}`));
    }

    if (options.exitOnGap && result.gaps.length > 0) {
      process.exit(2);
    }
  });

interface SchemaGap {
  path: string;
  method: string;
  operationId: string;
  category: string;
  detail: string;
}

File: rust/src/cli/validate_schemas.rs
//! `arbiter validate-schemas` (port of src/commands/validate-schemas.ts).

use std::path::Path;

use clap::Args;

use super::fail;
use crate::diff::{validate_schema_coverage, SchemaGap};

/// Validate schema coverage in an OpenAPI spec
#[derive(Args)]
pub struct ValidateSchemasArgs {
    /// path to OpenAPI spec
    #[arg(short = 's', long = "spec")]
    pub spec: String,

    /// path to write JSON validation report
    #[arg(short = 'o', long = "output")]
    pub output: Option<String>,

    /// exit with code 2 if gaps are found
    #[arg(long = "exit-on-gap")]
    pub exit_on_gap: bool,
}

pub fn run(args: &ValidateSchemasArgs) -> i32 {
    let result = match validate_schema_coverage(Path::new(&args.spec)) {
        Ok(result) => result,
        Err(e) => fail(format!("Schema validation failed: {e}")),
    };

    println!("\nSchema Coverage Report:");
    println!("  Total endpoints: {}", result.summary.total_endpoints);
    println!(
        "  Missing response schemas: {}",
        count_label(result.summary.missing_response_schemas)
    );
    println!(
        "  Missing request schemas: {}",
        count_label(result.summary.missing_request_schemas)
    );
    println!(
        "  Bare response schemas: {}",
        count_label_bare(result.summary.bare_response_schemas)
    );
    println!(
        "  Missing parameter schemas: {}",
        count_label(result.summary.missing_param_schemas)
    );
    println!("  Total gaps: {}", count_label(result.summary.total_gaps));

    if result.gaps.is_empty() {
        println!("\n✓ All endpoints have complete schema coverage!");
    } else {
        for (category, gaps) in group_by_category(&result.gaps) {
            println!("\n{category}:");
            for gap in gaps {
                println!("  {} {} | {}", gap.method, gap.path, gap.operation_id);
                println!("    → {}", gap.detail);
            }
        }
    }

    if let Some(output) = args.output.as_deref() {
        if let Err(e) = crate::diff::write_schema_validation_report(&result, Path::new(output)) {
            fail(format!("Failed to write report {output}: {e}"));
        }
        println!("\nReport written to {output}");
    }

    if args.exit_on_gap && !result.gaps.is_empty() {
        return 2;
    }
    0
}

fn count_label(count: usize) -> String {
    if count > 0 {
        count.to_string()
    } else {
        "0".to_string()
    }
}

/// TS colors bare response schema counts yellow rather than red; the plain
/// rendering is identical either way.
fn count_label_bare(count: usize) -> String {
    count_label(count)
}

/// Group gaps by serialized category, preserving first-seen category order
/// (mirrors the Map insertion order in TS).
fn group_by_category(gaps: &[SchemaGap]) -> Vec<(String, Vec<&SchemaGap>)> {
    let mut order: Vec<String> = Vec::new();
    let mut groups: std::collections::HashMap<String, Vec<&SchemaGap>> =
        std::collections::HashMap::new();
    for gap in gaps {
        let category = serde_json::to_value(gap)
            .ok()
            .and_then(|v| v.get("category").and_then(|c| c.as_str()).map(String::from))
            .unwrap_or_default();
        if !groups.contains_key(&category) {
            order.push(category.clone());
        }
        groups.entry(category).or_default().push(gap);
    }
    order
        .into_iter()
        .filter_map(|category| groups.remove_entry(&category))
        .collect()
}
